core/unit_test/incremental_testing/Test02c_MemorySpace.hpp

```diff
@@ -0,0 +1,114 @@
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <cstdio>
+#include <sstream>
+#include <type_traits>
+#include <gtest/gtest.h>
+
+namespace Test {
+
+// DeepCopy unit tests
+// We allocate data in the host. Copy it to the device and copy back the data
+// from device back to host We check if the original and copied data are the
+// same to evaluate the deep copies.
+
+template <class MemSpaceD, class MemSpaceH>
+struct TestIncrMemorySpace_deepcopy {
+  using dataType         = double;
+  const int num_elements = 10;
+  const double value     = 0.5;
+
+  int compare_equal_host(dataType *HostData_send, dataType *HostData_recv) {
+    int error = 0;
+    for (int i = 0; i < num_elements; ++i) {
+      if (HostData_send[i] != HostData_recv[i]) error++;
+    }
+
+    return error;
+  }
+
+  void testit_DtoH() {
+    // Allocate memory on Device space
+    dataType *DeviceData = (dataType *)Kokkos::kokkos_malloc<MemSpaceD>(
+        "DeviceData", num_elements * sizeof(dataType));
+    ASSERT_FALSE(DeviceData == nullptr);
+
+    // Allocate memory on Host space
+    dataType *HostData_send = (dataType *)Kokkos::kokkos_malloc<MemSpaceH>(
+        "HostData", num_elements * sizeof(dataType));
+    ASSERT_FALSE(HostData_send == nullptr);
+
+    // Allocate memory on Host space
+    dataType *HostData_recv = (dataType *)Kokkos::kokkos_malloc<MemSpaceH>(
+        "HostData", num_elements * sizeof(dataType));
+    ASSERT_FALSE(HostData_recv == nullptr);
+
+    for (int i = 0; i < num_elements; ++i) {
+      HostData_send[i] = value;
```

Comment:
fill there with different values for each element (something like `value + i` or so).